nexus/types/src/deployment.rs

```diff
@@ -271,31 +300,231 @@ impl<'a> fmt::Display for BlueprintDisplay<'a> {
+            width = BlueprintZoneDisposition::DISPLAY_WIDTH,
+        )
+    }
+}
+
+/// The desired state of an Omicron-managed zone in a blueprint.
+///
+/// Part of [`BlueprintZoneConfig`].
+#[derive(
+    Debug,
+    Copy,
+    Clone,
+    PartialEq,
+    Eq,
+    Hash,
+    PartialOrd,
+    Ord,
+    JsonSchema,
+    Deserialize,
+    Serialize,
+    EnumIter,
+)]
+#[serde(rename_all = "snake_case")]
+pub enum BlueprintZoneDisposition {
+    /// The zone is in-service.
+    InService,
+
+    /// The zone is not in service.
+    Quiesced,
+}
+
+impl BlueprintZoneDisposition {
+    /// The maximum width of `Display` output.
+    const DISPLAY_WIDTH: usize = 10;
+
+    /// Returns true if the zone disposition matches this filter.
+    pub fn matches(self, filter: BlueprintZoneFilter) -> bool {
+        // This code could be written in three ways:
+        //
+        // 1. match self { match filter { ... } }
+        // 2. match filter { match self { ... } }
+        // 3. match (self, filter) { ... }
+        //
+        // We choose 1 here because we expect many filters and just a few
+        // dispositions, and 1 is the easiest form to represent that.
+        match self {
+            Self::InService => match filter {
+                BlueprintZoneFilter::All => true,
+                BlueprintZoneFilter::InternalDns => true,
+                BlueprintZoneFilter::VpcFirewall => true,
+            },
+            Self::Quiesced => match filter {
+                BlueprintZoneFilter::All => true,
+
+                // Quiesced zones should not be exposed in DNS.
+                BlueprintZoneFilter::InternalDns => false,
+
+                // Quiesced zones should get firewall rules.
+                BlueprintZoneFilter::VpcFirewall => true,
+            },
+        }
+    }
+
+    /// Returns all zone dispositions that match the given filter.
+    pub fn all_matching(
+        filter: BlueprintZoneFilter,
+    ) -> impl Iterator<Item = Self> {
+        BlueprintZoneDisposition::iter().filter(move |&d| d.matches(filter))
+    }
+}
+
+impl fmt::Display for BlueprintZoneDisposition {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            // Neither `write!(f, "...")` nor `f.write_str("...")` obey fill
+            // and alignment (used above), but this does.
+            BlueprintZoneDisposition::InService => "in service".fmt(f),
+            BlueprintZoneDisposition::Quiesced => "quiesced".fmt(f),
+        }
+    }
+}
+
+/// Filters that apply to blueprint zones.
+///
+/// This logic lives here rather than within the individual components making
+/// decisions, so that this is easier to read.
+///
+/// The meaning of a particular filter should not be overloaded -- each time a
+/// new use case wants to make a decision based on the zone disposition, a new
+/// variant should be added to this enum.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum BlueprintZoneFilter {
```

Comment:
Making this a separate type is interesting; I don't think I have strong feelings about it either way, but it's not what I expected. I thought we'd have methods on `BlueprintZoneDisposition` itself like `.in_internal_dns()`, `.needs_firewall_rules()`, etc., and that the filter we pass to `all_blueprint_zones` would be a `BlueprintZoneDisposition -> bool` closure that we'd expect to be a call to one of those methods. By taking an explicit filter, though, it does prevent callers from doing something goofy in that closure, and maybe forces them to add a specific filter if one of the existing ones isn't right for them?

Reply:
The separate type was not actually  the first implementation, but I realized it provided a really nice orthogonality to the API. With just a single implementation, we got 3/4 sets of methods:

* matches
* all_matching
* all_blueprint_zones
* and in the future, all_omicron_zones

You could do this with a callback, but the explicit filter makes it so that new callers have to add their condition here rather than doing something bespoke in the caller (e.g. == in-service).